libflux/src/semantic/types.rs

```diff
@@ -116,27 +205,251 @@ impl fmt::Display for MonoType {
+    fn satisfies_at_most(
+        self,
+        kinds: HashSet<Kind>,
+    ) -> impl FnOnce(HashSet<Kind>, Const) -> Result {
+        move |k, c| {
+            if k.is_subset(&kinds) {
+                Ok((Subst::empty(), c))
+            } else {
+                Err(Error::Constrain {
+                    t: self,
+                    with: k.difference(&kinds).map(|&kind| kind).collect(),
+                })
+            }
+        }
+    }
+
+    fn contains(&self, tv: Tvar) -> bool {
+        match self {
+            MonoType::Bool
+            | MonoType::Int
+            | MonoType::Uint
+            | MonoType::Float
+            | MonoType::String
+            | MonoType::Duration
+            | MonoType::Time
+            | MonoType::Regexp => false,
+            MonoType::Var(tvr) => tv == *tvr,
+            MonoType::Arr(arr) => arr.contains(tv),
+            MonoType::Row(row) => row.contains(tv),
+            MonoType::Fun(fun) => fun.contains(tv),
+        }
+    }
+}
+
 // Tvar stands for type variable.
 // A type variable holds an unknown type.
-#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
-pub struct Tvar(i64);
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct Tvar(pub i64);
 
 impl fmt::Display for Tvar {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         write!(f, "t{}", self.0)
     }
 }
 
+impl Tvar {
+    fn apply(self, sub: &Subst) -> MonoType {
+        match sub.lookup(self) {
+            Some(t) => t.clone(),
+            None => MonoType::Var(self),
+        }
+    }
+
+    fn unify(self, with: MonoType, cons: Const) -> Result {
+        match with {
+            MonoType::Var(tv) => {
+                if self == tv {
+                    // The empty substitution will always
+                    // unify a type variable with itself.
+                    Ok((Subst::empty(), cons))
+                } else {
+                    // Unify two distinct type variables.
+                    // This will update the kind constraints
+                    // associated with these type variables.
+                    self.unify_with_tvar(tv, cons)
+                }
+            }
+            _ => {
+                if with.contains(self) {
+                    // Invalid recursive type
+                    Err(Error::Occurs { tv: self, t: with })
+                } else {
+                    // Unify a type variable with a monotype.
+                    // The monotype must satisify any
+                    // constraints placed on the type variable.
+                    self.unify_with_type(with, cons)
+                }
+            }
+        }
+    }
+
+    fn unify_with_tvar(self, tv: Tvar, cons: Const) -> Result {
+        let mut cons = cons;
+        // Gather the kind constraints for both type variables
+        let kinds: HashSet<Kind> = cons
+            .0
+            .get(&self)
+            .unwrap_or(&HashSet::new())
```

Comment:
nit, in reading up on what this does, I discovered that `unwrap_or` is evaluated eagerly so we are creating a new hashset with each call to `unify_with_tvar`. The docs recommend using `unwrap_or_else` which accepts a closure which is evaluated lazily.

Reply:
I must have missed `unwrap_or_else`. I will update. Note however that while this creates a new hashset value, it won't actually allocate any heap memory until we insert the first element.

Reply:
That is good to know, I see that in the docs now. In that case i say this reads more clearly so we should leave it as `unwrap_or` since we do not need to avoid the allocation.